Use the Password_Checker package

briancolinger opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes #
Depends on #18700.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Uses the Password_Checker package.
* Removes the usage of jetpack_require_lib().

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Open wp shell for a Jetpack site. 
* `yarn docker:wp shell --url=DOMAIN`
* wp_set_current_user( 1 );
* $um = Automattic\Jetpack\Sync\Modules::get_module( 'users' );
* $um->authenticate_handler( wp_get_current_user(), wp_get_current_user()->user_login, 'test' );
* Verify WP_User is returned.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* Sync: Use the new Password_Checker package instead of Jetpack_Password_Checker.

mdbitz commented on projects/packages/sync/src/modules/class-users.php:
```diff
@@ -8,7 +8,9 @@
 use Automattic\Jetpack\Constants as Jetpack_Constants;
+use Automattic\Jetpack\Password_Checker;
 use Automattic\Jetpack\Sync\Defaults;
+namespace Automattic\Jetpack;
```
This namespace line needs to be removed.

briancolinger replied:
Fixed.

mdbitz reviewed (changes requested):
There is an errant namespace statement that needs to be removed.

Beyond that the description of the PR should reference the dependency on #18700 (as in this PR will be re-based after 18700 is merged)

Also the test instructions should include steps that would trigger the authenticate_handler  function and confirm existing functionality is maintained.

briancolinger commented:
Looks like I committed the password-checker package here by mistake. Fixed that too.

briancolinger commented:
Fixed the problem that led me into dependency hell earlier today.
The authenticate_handler() method will now return a WP_Error() when the password test fails.

I also updated the testing instructions to include setting the current user_id to 1 and to pass the user_login from the WP_User object as the username parameter.

mdbitz commented on projects/packages/sync/src/modules/class-users.php:
```diff
@@ -380,7 +380,11 @@ public function authenticate_handler( $user, $username, $password ) {
 		);
 
-		return $user;
+		return new \WP_Error(
```
@briancolinger we can't modify this as it changes the current functionality which we don't want at this time.

Please revert so it still returns $user always.

Instead we need to update our testing rules so that we validate if the flags (add_flags) were updated.

mdbitz replied:
apologies, i didn't read the function in full and was simply going off the doc comments :)

briancolinger replied:
Could you please clarify this for me?
Should I revert this change?

I thought it would be ok/consistent with the existing documented @return \WP_Error|\WP_User

zinigor replied:
Yes, I agree that we should revert to previous behavior, unless there's a good reason we should interrupt the authentication process. Before we were returning whatever we had been passed even when a problem was found because we didn't want to interrupt the user - we just wanted to set the flags that would get synced to the cache site and possibly then warn the administrator.

briancolinger replied:
Fixed.

briancolinger commented:
Ok, I think this thing is ready for review now.

zinigor commented:
It looks like tests are failing in Travis environment:
```
  1) WP_Test_Jetpack_Sync_Listener::test_does_listener_add_actor_user_data_for_login_events
  Error: Class 'Automattic\Jetpack\Password_Checker' not found
  
  /tmp/wordpress-latest/src/wp-content/plugins/jetpack/vendor/automattic/jetpack-sync/src/modules/class-users.php:366
  /tmp/wordpress-latest/src/wp-includes/class-wp-hook.php:292
  /tmp/wordpress-latest/src/wp-includes/plugin.php:212
  /tmp/wordpress-latest/src/wp-includes/pluggable.php:582
  /tmp/wordpress-latest/src/wp-includes/user.php:95
  /tmp/wordpress-latest/src/wp-content/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-listener.php:127
```
Perhaps there's something that is there on your machine that is missing when a clean installation is performed?

kbrown9 reviewed (approved):
This worked for me!

I needed to manually add the jetpack-password-checker package to Jetpack's `composer.json` file in my local repo after checking out this branch. I'm not sure why the change to that file wasn't applied when I checked out the branch.

Also, the testing instructions state that `authenticate_handler` will return a `WP_Error` object if the tests fail. In my tests, the method returned a `WP_User` object when the tests failed. I think that's expected based on [this](https://github.com/Automattic/jetpack/blob/master/projects/packages/sync/src/modules/class-users.php#L383).

briancolinger commented:
> returned a `WP_User` object

Ah, my bad! That was true before I reverted the change that made the method return the WP_Error. I reverted the change to avoid breaking backwards compatibility.
I've updated the testing instructions.

zinigor reviewed (approved):
Let's merge this, thanks for the work on this, Brian!
